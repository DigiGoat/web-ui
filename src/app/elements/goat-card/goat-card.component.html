<div class="card color-scheme-secondary" *ngIf="goat" test-id="goat">
  <div class="px-3 pt-3">
    <img *ngIf="image?.file" [ngSrc]="image!.file" [alt]="image?.alt" class="card-img-top rounded object-fit-contain" width="400"
      height="250" test-id="goat-image">
  </div>
  <div class="card-body color-scheme-secondary" [id]="identifier">
    <div class="container color-scheme-tertiary pb-3">
      <a *ngIf="identifier" class="stretched-link" [routerLink]="identifier" test-id="goat-link"
        attr.aria-label="Link to open {{identifier}}'s modal (This link cover's the whole element - it does not contain any text)"></a>
      <h5 class="card-title text-center mb-0" test-id="goat-name">{{name}}</h5>
      <h6 class="card-subtitle pb-1 fw-light fst-italic text-center" test-id="goat-age" *ngIf="born">{{born | age}}</h6>
      <h6 class="card-subtitle pb-1 text-center fw-light fst-italic font-monospace" *ngIf="linearAppraisal" test-id="goat-appraisal">
        {{linearAppraisal.generalAppearance}}{{linearAppraisal.dairyStrength}}{{linearAppraisal.bodyCapacity}}{{linearAppraisal.mammarySystem}}
        {{linearAppraisal.finalScore}}
      </h6>
      <h6 class="card-subtitle pb-1 text-center" *ngIf="nickname" test-id="goat-nickname">"{{nickname}}"</h6>
      <div class="card-text text-center text-truncate-3 pt-1" *ngIf="description" [innerHTML]="description" test-id="goat-description">
      </div>
    </div>
  </div>
</div>
<div class="card color-scheme-secondary" *ngIf="!goat" test-id="placeholder">
  <div class="px-3 pt-3 placeholder-glow">
    <img class="card-img-top rounded border placeholder" alt="Loading..." width="400" height="200" test-id="placeholder-image">
  </div>
  <div class="card-body color-scheme-secondary">
    <div class="container color-scheme-tertiary pb-3 text-center placeholder-wave">
      <h5 class="card-title text-center placeholder" test-id="placeholder-name">SOME ANIMALS NAME</h5>
      <br>
      <h6 class="card-subtitle pb-2 text-center placeholder" test-id="placeholder-nickname">"Nickname"</h6>
    </div>
  </div>
</div>