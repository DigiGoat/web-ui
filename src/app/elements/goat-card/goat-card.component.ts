import { Component, Input, OnInit } from '@angular/core';

import { ImageEntry, ImageService } from '../../services/image/image.service';


import type { Goat } from 'src/app/services/goat/goat.service';
@Component({
  selector: 'app-goat-card',
  templateUrl: './goat-card.component.html',
  styleUrls: ['./goat-card.component.scss']
})
export class GoatCardComponent implements OnInit {
  constructor(private imageService: ImageService) { }

  @Input() goat?: Partial<Goat>;
  name?: string;
  nickname?: string;
  description?: string;
  id?: string;
  born?: string;
  image?: ImageEntry;
  identifier?: string;
  ngOnInit(): void {
    this.name = this.goat?.name;
    this.nickname = this.goat?.nickname;
    this.description = this.goat?.description;
    this.id = this.goat?.normalizeId;
    this.born = this.goat?.dateOfBirth;
    this.image = this.imageService.getImage([this.id, this.name, this.nickname]);
    this.identifier = this.nickname ?? this.id ?? this.name;
  }
}
