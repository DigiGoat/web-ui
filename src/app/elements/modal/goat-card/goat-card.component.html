<h5 class="card-title text-center" *ngIf="name" test-id="name">{{name}}</h5>
<h6 class="card-subtitle mb-2 text-center" *ngIf="nickname" test-id="nickname">"{{nickname}}"</h6>
<div class="card-header color-scheme-quaternary">
  <app-carousel [images]="images" test-id="carousel"></app-carousel>
  <div class="card-text text-center container color-scheme-tertiary rounded-bottom">
    <div class="text-capitalize">
      <div *ngIf="born">
        <span class="fw-bolder">Born</span>: <span class="fw-lighter" test-id="born">{{born | date:'longDate'}}</span>
        <br>
      </div>
      <div *ngIf="id">
        <span class="fw-bolder">ID</span>: <span class="fw-lighter" test-id="id">{{id}}</span>
        <br>
      </div>
      <div *ngIf="animalTattoos?.length">
        <span class="fw-bolder">Tattoo</span>: <span *ngFor="let animalTattoo of animalTattoos" test-id="tattoo">
          <span *ngIf="animalTattoo.tattooLocation?.name && animalTattoo.tattoo">
            {{animalTattoo.tattooLocation?.name}}:<span class="fw-lighter">{{animalTattoo.tattoo}}</span>
          </span>
        </span>
        <br>
      </div>
      <div *ngIf="colorAndMarking">
        <span class="fw-bolder">Markings</span>: <span class="fw-lighter" test-id="markings">{{colorAndMarking.toLowerCase()}}</span>
      </div>
    </div>
    <span test-id="description" *ngIf="description">{{description}}</span>
  </div>
</div>
