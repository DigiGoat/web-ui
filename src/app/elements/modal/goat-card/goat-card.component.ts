import { Component, Input, type OnInit } from '@angular/core';
import { Meta } from '@angular/platform-browser';
import type { Observable } from 'rxjs';
import type { Page } from '../../../app-routing.module';
import type { Goat } from '../../../services/goat/goat.service';
import { ImageService, type ImageEntry } from '../../../services/image/image.service';

@Component({
  selector: 'app-modal-goat-card',
  templateUrl: './goat-card.component.html',
  styleUrl: './goat-card.component.scss'
})
export class GoatCardComponent implements OnInit, Page {
  @Input({ required: true }) goat?: Partial<Goat>;

  constructor(public imageService: ImageService, private meta: Meta) { }
  setDescription(): void | Observable<void> {
    if (this.description) {
      this.meta.addTags([{ name: 'og:description', content: this.description }, { name: 'description', content: this.description }]);
    }
  }

  name?: string;
  nickname?: string;
  description?: string;
  id?: string;
  born?: string | number;
  animalTattoos?: Goat['animalTattoo'];
  colorAndMarking?: string;
  images!: ImageEntry[];
  ngOnInit(): void {
    this.name = this.goat?.name;
    this.nickname = this.goat?.nickname;
    this.description = this.goat?.description;
    this.id = this.goat?.normalizeId;
    this.born = this.goat?.dateOfBirth;
    this.animalTattoos = this.goat?.animalTattoo;
    this.colorAndMarking = this.goat?.colorAndMarking;
    this.images = this.imageService.getImages([this.id, this.name, this.nickname]);
    this.setDescription();
  }
}
