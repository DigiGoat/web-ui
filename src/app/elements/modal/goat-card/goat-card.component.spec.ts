import { ComponentFixture, TestBed } from '@angular/core/testing';

import type { Meta } from '@angular/platform-browser';
import { GoatCardComponent } from './goat-card.component';

describe('GoatCardComponent', () => {
  let component: GoatCardComponent;
  let fixture: ComponentFixture<GoatCardComponent>;
  let html: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [GoatCardComponent]
    })
      .compileComponents();

    fixture = TestBed.createComponent(GoatCardComponent);
    component = fixture.componentInstance;
    html = fixture.nativeElement;
    component['meta'] = { addTags: jest.fn() } as unknown as Meta;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  describe('With an Empty Goat', () => {
    beforeEach(() => {
      component.goat = {};
      fixture.detectChanges();
    });
    it('should not set a page description', () => {
      expect(component['meta'].addTags).toHaveBeenCalledTimes(0);
    });
    it('should not display a name', () => {
      const element = html.querySelector('[test-id=name]');
      expect(element).toBeFalsy();
    });
    it('should not display a nickname', () => {
      const element = html.querySelector('[test-id=nickname]');
      expect(element).toBeFalsy();
    });
    it('should display carousel', () => {
      const element = html.querySelectorAll('[test-id=carousel]');
      expect(element).toHaveLength(1);
    });
    it('should not display a date of birth', () => {
      const element = html.querySelector('[test-id=born]');
      expect(element).toBeFalsy();
    });
    it('should not display an id', () => {
      const element = html.querySelector('[test-id=id]');
      expect(element).toBeFalsy();
    });
    it('should not display tattoos', () => {
      const element = html.querySelector('[test-id=tattoo]');
      expect(element).toBeFalsy();
    });
    it('should not display markings', () => {
      const element = html.querySelector('[test-id=markings]');
      expect(element).toBeFalsy();
    });
    it('should not display a description', () => {
      const element = html.querySelector('[test-id=description]');
      expect(element).toBeFalsy();
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
  describe('With a Goat', () => {
    beforeEach(() => {
      component.goat = {
        name: 'TEST_NAME',
        nickname: 'TEST_NICKNAME',
        description: 'TEST_DESCRIPTION',
        normalizeId: 'TEST_ID',
        dateOfBirth: '5/8/2024',
        animalTattoo: [{
          tattoo: 'TEST_TATTOO_1',
          tattooLocation: {
            name: 'TEST_LOCATION_1'
          }
        }, {
          tattoo: 'TEST_TATTOO_2',
          tattooLocation: {
            name: 'TEST_LOCATION_2'
          }
        }],
        colorAndMarking: 'TEST_COLOR_AND_MARKING'
      };
      fixture.detectChanges();
    });
    it('should set a page description', () => {
      expect(component['meta'].addTags).toHaveBeenCalledTimes(1);
      expect(component['meta'].addTags).toHaveBeenCalledWith([{ 'content': 'TEST_DESCRIPTION', 'name': 'og:description' }, { 'content': 'TEST_DESCRIPTION', 'name': 'description' }]);
    });
    it('should display a name', () => {
      const element = html.querySelector('[test-id=name]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_NAME');
    });
    it('should display a nickname', () => {
      const element = html.querySelector('[test-id=nickname]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('"TEST_NICKNAME"');
    });
    it('should display carousel', () => {
      const element = html.querySelectorAll('[test-id=carousel]');
      expect(element).toHaveLength(1);
    });
    it('should display a date of birth', () => {
      const element = html.querySelector('[test-id=born]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('May 8, 2024');
    });
    it('should display an id', () => {
      const element = html.querySelector('[test-id=id]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_ID');
    });
    it('should display tattoos', () => {
      const element = html.querySelectorAll('[test-id=tattoo]');
      expect(element).toHaveLength(2);
      expect(element[0].innerHTML).toContain('<span> TEST_LOCATION_1:<span class="fw-lighter">TEST_TATTOO_1</span></span>');
      expect(element[1].innerHTML).toContain('<span> TEST_LOCATION_2:<span class="fw-lighter">TEST_TATTOO_2</span></span>');
    });
    it('should display markings', () => {
      const element = html.querySelector('[test-id=markings]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('test_color_and_marking');
    });
    it('should display a description', () => {
      const element = html.querySelector('[test-id=description]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_DESCRIPTION');
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
});
