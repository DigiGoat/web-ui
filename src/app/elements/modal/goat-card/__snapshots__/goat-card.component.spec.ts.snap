// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoatCardComponent With a Goat should match the snapshot 1`] = `
<app-modal-goat-card
  animalTattoos={[Function Array]}
  born={[Function String]}
  colorAndMarking={[Function String]}
  description={[Function String]}
  goat={[Function Object]}
  id={[Function String]}
  imageService={[Function ImageService]}
  images={[Function Array]}
  meta={[Function Object]}
  name={[Function String]}
  nickname={[Function String]}
>
  <h5
    class="card-title text-center"
    test-id="name"
  >
    TEST_NAME
  </h5><h6
    class="card-subtitle mb-2 text-center"
    test-id="nickname"
  >
    "TEST_NICKNAME"
  </h6><div
    class="card-header color-scheme-quaternary"
  >
    <app-carousel
      test-id="carousel"
    />
    <div
      class="card-text text-center container color-scheme-tertiary rounded-bottom"
    >
      <div
        class="text-capitalize"
      >
        <div>
          <span
            class="fw-bolder"
          >
            Born
          </span>
          : 
          <span
            class="fw-lighter"
            test-id="born"
          >
            May 8, 2024
          </span>
          <br />
        </div>
        <div>
          <span
            class="fw-bolder"
          >
            ID
          </span>
          : 
          <span
            class="fw-lighter"
            test-id="id"
          >
            TEST_ID
          </span>
          <br />
        </div>
        <div>
          <span
            class="fw-bolder"
          >
            Tattoo
          </span>
          : 
          <span
            test-id="tattoo"
          >
            <span>
               TEST_LOCATION_1:
              <span
                class="fw-lighter"
              >
                TEST_TATTOO_1
              </span>
            </span>
          </span>
          <span
            test-id="tattoo"
          >
            <span>
               TEST_LOCATION_2:
              <span
                class="fw-lighter"
              >
                TEST_TATTOO_2
              </span>
            </span>
          </span>
          <br />
        </div>
        <div>
          <span
            class="fw-bolder"
          >
            Markings
          </span>
          : 
          <span
            class="fw-lighter"
            test-id="markings"
          >
            test_color_and_marking
          </span>
        </div>
      </div>
      <span
        test-id="description"
      >
        TEST_DESCRIPTION
      </span>
    </div>
  </div>
</app-modal-goat-card>
`;

exports[`GoatCardComponent With an Empty Goat should match the snapshot 1`] = `
<app-modal-goat-card
  animalTattoos="undefined"
  born="undefined"
  colorAndMarking="undefined"
  description="undefined"
  goat={[Function Object]}
  id="undefined"
  imageService={[Function ImageService]}
  images={[Function Array]}
  meta={[Function Object]}
  name="undefined"
  nickname="undefined"
>
  <div
    class="card-header color-scheme-quaternary"
  >
    <app-carousel
      test-id="carousel"
    />
    <div
      class="card-text text-center container color-scheme-tertiary rounded-bottom"
    >
      <div
        class="text-capitalize"
      >
      </div>
    </div>
  </div>
</app-modal-goat-card>
`;
